File: eval/replay_harness/event_time_buffer.py
"""Deterministic event-time buffering (CURIE-006).

Policy
------
- Buffer arrivals until the event-time watermark advances (``max_event_time - allowed_lateness``)
  or ``close()`` is called.
- Flush in order of ``(event_time, tie_breaker)`` so Kafka arrival order cannot change results
  within the lateness window.
- Events with ``event_time < watermark`` (beyond allowed lateness) are **not** applied to
  feature or governance state; they are returned with disposition ``late_beyond_lateness``.
- Late corrections do **not** retract prior alerts (future-state only).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Generic, TypeVar

T = TypeVar("T")


@dataclass(frozen=True, order=True)
class _SortKey:
    event_time_ms: int
    tie_breaker: str


@dataclass
class BufferedEvent(Generic[T]):
    event_time_ms: int
    tie_breaker: str
    payload: T


@dataclass
class FlushResult(Generic[T]):
    """Events ready for state mutation, plus late drops for audit/DLQ."""

    ready: list[BufferedEvent[T]] = field(default_factory=list)
    late: list[BufferedEvent[T]] = field(default_factory=list)
    disposition_late: str = "late_beyond_lateness"


class EventTimeBuffer(Generic[T]):
    """Per-key event-time reorder buffer."""

    def __init__(self, *, allowed_lateness_ms: int = 5 * 60 * 1000) -> None:
        if allowed_lateness_ms < 0:
            raise ValueError("allowed_lateness_ms must be >= 0")
        self.allowed_lateness_ms = allowed_lateness_ms
        self._pending: list[BufferedEvent[T]] = []
        self.max_event_time_ms: int | None = None
        self.watermark_ms: int = 0

    def offer(
        self,
        event_time_ms: int,
        payload: T,
        *,
        tie_breaker: str = "",
    ) -> FlushResult[T]:
        """Admit an event; may flush previously buffered events as the watermark advances."""
        if self.max_event_time_ms is None or event_time_ms > self.max_event_time_ms:
            self.max_event_time_ms = event_time_ms
            self.watermark_ms = max(0, event_time_ms - self.allowed_lateness_ms)

        if event_time_ms < self.watermark_ms:
            return FlushResult(late=[BufferedEvent(event_time_ms, tie_breaker, payload)])

        self._pending.append(BufferedEvent(event_time_ms, tie_breaker, payload))
        return self._flush_ready()

    def close(self) -> FlushResult[T]:
        """End-of-stream: release everything remaining in order."""
        self.watermark_ms = (
            self.max_event_time_ms + 1 if self.max_event_time_ms is not None else 0
        )
        return self._flush_ready()

    def snapshot_pending(self) -> list[BufferedEvent[T]]:
        """Restart-friendly copy of buffered (not yet flushed) events."""
        return list(self._pending)

    def restore_pending(self, events: list[BufferedEvent[T]]) -> None:
        self._pending = list(events)
        if events:
            self.max_event_time_ms = max(e.event_time_ms for e in events)
            self.watermark_ms = max(0, self.max_event_time_ms - self.allowed_lateness_ms)

    def _flush_ready(self) -> FlushResult[T]:
        ready: list[BufferedEvent[T]] = []
        stay: list[BufferedEvent[T]] = []
        for ev in self._pending:
            if ev.event_time_ms <= self.watermark_ms:
                ready.append(ev)
            else:
                stay.append(ev)
        self._pending = stay
        ready.sort(key=lambda e: _SortKey(e.event_time_ms, e.tie_breaker))
        return FlushResult(ready=ready)

File: streaming/flink-jobs/sofa/src/main/java/com/curie/sofa/state/EventTimeBuffer.java
package com.curie.sofa.state;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * Deterministic event-time reorder buffer (CURIE-006 / CURIE-026).
 *
 * <p>Within {@code allowedLatenessMs}, arrivals are buffered and flushed in {@code (eventTimeMs,
 * tieBreaker)} order. Beyond lateness: disposition {@code late_beyond_lateness} — do not mutate
 * feature/governance state; late corrections do not retract prior alerts.
 *
 * <p>A single event is not flushed from {@link #offer} alone: watermark starts at {@code eventTime -
 * lateness}. Operators must register an event-time timer at {@link #flushTimerTimestamp(long)} (or
 * call {@link #advanceWatermark(long)} / {@link #close()}) so the final event for a key still
 * scores without a following arrival.
 */
public final class EventTimeBuffer<T> implements Serializable {
  public static final String LATE_DISPOSITION = "late_beyond_lateness";

  /** Shared SOFA/AKI event-time policy version (CURIE-026 / CURIE-027). */
  public static final String POLICY_VERSION = "event-time-v1-lateness-5m";

  public static final class BufferedEvent<T> implements Serializable {
    public final long eventTimeMs;
    public final String tieBreaker;
    public final T payload;

    public BufferedEvent(long eventTimeMs, String tieBreaker, T payload) {
      this.eventTimeMs = eventTimeMs;
      this.tieBreaker = tieBreaker != null ? tieBreaker : "";
      this.payload = payload;
    }
  }

  public static final class FlushResult<T> implements Serializable {
    public final List<BufferedEvent<T>> ready = new ArrayList<>();
    public final List<BufferedEvent<T>> late = new ArrayList<>();
    public final String dispositionLate = LATE_DISPOSITION;
  }

  private final long allowedLatenessMs;
  private final List<BufferedEvent<T>> pending = new ArrayList<>();
  private Long maxEventTimeMs;
  private long watermarkMs;

  public EventTimeBuffer(long allowedLatenessMs) {
    if (allowedLatenessMs < 0) {
      throw new IllegalArgumentException("allowedLatenessMs must be >= 0");
    }
    this.allowedLatenessMs = allowedLatenessMs;
  }

  public long allowedLatenessMs() {
    return allowedLatenessMs;
  }

  public long watermarkMs() {
    return watermarkMs;
  }

  public Long maxEventTimeMs() {
    return maxEventTimeMs;
  }

  public int pendingCount() {
    return pending.size();
  }

  public String policyVersion() {
    return POLICY_VERSION;
  }

  /**
   * Event-time timer timestamp that advances the local watermark far enough to flush {@code
   * eventTimeMs} (and any earlier pending events).
   */
  public long flushTimerTimestamp(long eventTimeMs) {
    return eventTimeMs + allowedLatenessMs;
  }

  /**
   * Advance the local watermark (e.g. from an event-time timer) and flush eligible pending
   * events. Watermark only moves forward.
   */
  public FlushResult<T> advanceWatermark(long newWatermarkMs) {
    if (newWatermarkMs > watermarkMs) {
      watermarkMs = newWatermarkMs;
    }
    return flushReady();
  }

  public FlushResult<T> offer(long eventTimeMs, T payload, String tieBreaker) {
    if (maxEventTimeMs == null || eventTimeMs > maxEventTimeMs) {
      maxEventTimeMs = eventTimeMs;
      watermarkMs = Math.max(0L, eventTimeMs - allowedLatenessMs);
    }
    FlushResult<T> out = new FlushResult<>();
    if (eventTimeMs < watermarkMs) {
      out.late.add(new BufferedEvent<>(eventTimeMs, tieBreaker, payload));
      return out;
    }
    pending.add(new BufferedEvent<>(eventTimeMs, tieBreaker, payload));
    return flushReady();
  }

  public FlushResult<T> close() {
    watermarkMs = maxEventTimeMs != null ? maxEventTimeMs + 1 : 0L;
    return flushReady();
  }

  public List<BufferedEvent<T>> snapshotPending() {
    return new ArrayList<>(pending);
  }

  public void restorePending(List<BufferedEvent<T>> events) {
    pending.clear();
    pending.addAll(events);
    maxEventTimeMs = null;
    for (BufferedEvent<T> e : pending) {
      if (maxEventTimeMs == null || e.eventTimeMs > maxEventTimeMs) {
        maxEventTimeMs = e.eventTimeMs;
      }
    }
    watermarkMs =
        maxEventTimeMs == null ? 0L : Math.max(0L, maxEventTimeMs - allowedLatenessMs);
  }

  private FlushResult<T> flushReady() {
    FlushResult<T> out = new FlushResult<>();
    List<BufferedEvent<T>> stay = new ArrayList<>();
    for (BufferedEvent<T> ev : pending) {
      if (ev.eventTimeMs <= watermarkMs) {
        out.ready.add(ev);
      } else {
        stay.add(ev);
      }
    }
    pending.clear();
    pending.addAll(stay);
    out.ready.sort(
        Comparator.comparingLong((BufferedEvent<T> e) -> e.eventTimeMs)
            .thenComparing(e -> e.tieBreaker));
    return out;
  }
}
